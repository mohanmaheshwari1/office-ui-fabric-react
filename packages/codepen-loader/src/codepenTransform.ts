import * as babylon from 'babylon';
import * as prettier from 'prettier';
import * as fs from 'fs';
import * as path from 'path';

import { Recast, JSCodeShift, IASTPath, IASTSpecifier, IASTNode } from './interfaces';

// recast and jscodeshift have no @types :(
const recast: Recast = require('recast');
const jscodeshift: {
  withParser: (parser: string) => JSCodeShift;
} = require('jscodeshift');

// These files are copied to lib in the pre-copy step (see config/pre-copy.json)
const exampleDataFiles = {
  exampleData: { name: 'exampleData', contents: '' },
  testImages: { name: 'TestImages', contents: '' },
  peopleExampleData: { name: 'PeopleExampleData', contents: '' }
};
// Regex to match example data file imports
const exampleDataRegex = new RegExp(
  Object.values(exampleDataFiles)
    // tslint:disable-next-line:typedef
    .map(fileInfo => `\\b${fileInfo.name}\\b`)
    .join('|') + '$'
);

for (const fileInfo of Object.values(exampleDataFiles)) {
  // Read each file and remove all export statements (to avoid confusing the logic later that
  // determines which exported thing is the actual example component)
  let contents = fs
    .readFileSync(path.resolve(__dirname, '../lib', fileInfo.name + '.ts'))
    .toString();

  // Strip the first few lines of comments from the files--they all have notes at the top which
  // don't need to be in the resulting codepen
  while (contents.startsWith('//')) {
    contents = contents.replace(/^.*?\r?\n/, '');
  }
  // Hack to quickly get rid of exports
  contents = contents.replace(/^export /gm, '');
  fileInfo.contents = contents;
}

// PeopleExampleData imports TestImages and uses it in a constant.
// Replace the import with the TestImages file contents to avoid a runtime error.
exampleDataFiles.peopleExampleData.contents = exampleDataFiles.peopleExampleData.contents.replace(
  /import .*?\bTestImages\b.*?\r?\n/,
  exampleDataFiles.testImages.contents
);

const j = jscodeshift.withParser('babylon');

const parse = (source: string) =>
  babylon.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript', 'classProperties', 'objectRestSpread']
  });

/**
 * This Transform modifies Fabric website code examples into a format that will allow them to be
 * rendered on Codepen, as part of the "Export to Codepen" feature.
 * There are two types of supported example templates:
 *
 *     [imports]
 *     [named variable export with example code inside]
 *
 * and
 *
 *     [imports]
 *     [named class export with example code inside]
 *
 * Currently, examples which are scattered across multiple files are NOT supported.
 *
 * @param file Example file contents to transform
 * @returns The transformed file contents
 */
export function transform(file: string): string {
  let sourceStr = file;
  // If example data files were imported, append the file contents
  for (const fileInfo of Object.values(exampleDataFiles)) {
    if (file.includes(`/${fileInfo.name}'`)) {
      sourceStr += `\n${fileInfo.contents}\n`;
    }
  }

  const source = j(recast.parse(sourceStr, { parser: { parse } }));

  // Make a list of imported identifiers, and remove all imports
  const identifiers: string[] = [];
  source.find(j.ImportDeclaration).forEach((p: IASTPath) => {
    const importPath = p.node.source.value;
    // Ignore identifiers from:
    // - the React import (which will be a global)
    // - css/scss
    // - example data files which will be appended if needed
    if (
      importPath !== 'react' &&
      !/\.s?css$/.test(importPath) &&
      !exampleDataRegex.test(importPath)
    ) {
      p.node.specifiers.forEach((spec: IASTSpecifier) => {
        identifiers.push(spec.local.loc.identifierName);
      });
    }

    // Remove the import
    p.prune();
  });

  let exampleName;
  // remove exports and replace with variable or class declarations, whichever the original example used
  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) =>
        node.declaration.type === 'VariableDeclaration' ||
        node.declaration.type === 'FunctionDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'VariableDeclaration') {
        exampleName = p.node.declaration.declarations[0].id.name;
      }
      return p.node.declaration;
    });

  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) =>
        node.declaration.type === 'ClassDeclaration' ||
        node.declaration.type === 'TSInterfaceDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'ClassDeclaration') {
        exampleName = p.node.declaration.id.name;
      }
      return p.node.declaration;
    });

  // Build the list of imports from window.Fabric
  let attachedWindowString = 'const {';
  if (identifiers.length > 0) {
    attachedWindowString += identifiers.join(',') + ',';
  }
  attachedWindowString += 'Fabric} = window.Fabric;\n';

  // add imports and React render footer (with the component wrapped in a <Fabric> for styling)
  const sourceWithFooter = [
    attachedWindowString,
    source.toSource(),
    `ReactDOM.render(<Fabric><${exampleName}/></Fabric>, document.getElementById("content"));`
  ].join('\n');

  return prettier.format(sourceWithFooter, {
    parser: 'typescript',
    printWidth: 100,
    tabWidth: 2,
    singleQuote: true
  });
}
