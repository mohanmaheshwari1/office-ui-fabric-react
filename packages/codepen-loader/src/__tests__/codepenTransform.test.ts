import { transform } from '../codepenTransform';
import * as fs from 'fs';
import * as path from 'path';

describe('codepen transform', () => {
  function transformFile(file: string): string {
    const filename = path.resolve(__dirname, './examples/' + file);
    const fileContents = fs.readFileSync(filename).toString();
    return transform(fileContents);
  }

  it('handles examples with function components', () => {
    const result = transformFile('function.txt');
    expect(result).toMatchSnapshot();
  });

  it('handles examples with class components', () => {
    const result = transformFile('class.txt');
    expect(result).toMatchSnapshot();
  });

  it('handles examples importing exampleData', () => {
    // Don't do a snapshot test on this file to avoid being sensitive to inconsequential changes.
    // Instead, just check for the main expected interface/functions.
    const result = transformFile('exampleData.txt');
    expect(result).toMatch(/^interface IExampleItem/m);
    expect(result).toMatch(/^function createListItems/m);
    expect(result).toMatch(/^function createGroups/m);
    expect(result).toMatch(/^function lorem/m);
    expect(result).toMatch(/^function isGroupable/m);
  });

  it('handles examples importing TestImages', () => {
    let result = transformFile('testImages.txt');
    result = shortenTestImages(result);
    expect(result).toMatchSnapshot();
  });

  it('handles examples importing PeopleExampleData', () => {
    let result = transformFile('people.txt');
    // Shorten the output and make it less sensitive to (likely inconsequential) data changes
    result = result.replace(/^(const people.*? = \[)$[\s\S]*?^(\];)$/m, '$1$2');
    result = shortenTestImages(result);
    expect(result).toMatchSnapshot();
  });
});

function shortenTestImages(result: string): string {
  return result.replace(/^(const TestImages = \{)$[\s\S]*?^(\};)$/m, '$1$2');
}
