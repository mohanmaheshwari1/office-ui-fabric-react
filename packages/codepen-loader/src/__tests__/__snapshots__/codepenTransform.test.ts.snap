// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codepen transform handles examples importing PeopleExampleData 1`] = `
"const {
  IExtendedPersonaProps,
  SelectedPeopleList,
  ISelectedPeopleItemProps,
  ExtendedSelectedItem,
  PrimaryButton,
  Selection,
  Stack,
  PersonaPresence,
  IExtendedPersonaProps,
  Fabric
} = window.Fabric;

interface ISelectedPeopleListBasicExampleState {
  nextPersonIndex: number;
}

class SelectedPeopleListBasicExample extends React.Component<
  {},
  ISelectedPeopleListBasicExampleState
> {
  private _selectionList = React.createRef<SelectedPeopleList>();
  private _selection: Selection;

  constructor(props: {}) {
    super(props);

    this.state = {
      nextPersonIndex: 0
    };
    this._selection = new Selection({ onSelectionChanged: () => this.forceUpdate() });
  }

  public render(): JSX.Element {
    return (
      <div>
        <PrimaryButton
          text=\\"Add another item\\"
          onClick={this._onAddItemButtonClicked}
          disabled={this.state.nextPersonIndex >= people.length}
          styles={{ root: { display: 'block', marginBottom: 20 } }}
        />
        <Stack horizontal wrap>
          <SelectedPeopleList
            key=\\"normal\\"
            removeButtonAriaLabel=\\"Remove\\"
            defaultSelectedItems={[people[40]]}
            componentRef={this._selectionList}
            onCopyItems={this._onCopyItems}
            onExpandGroup={this._onExpandItem}
            copyMenuItemText=\\"Copy\\"
            removeMenuItemText=\\"Remove\\"
            selection={this._selection}
            onRenderItem={this._onRenderItem}
          />
        </Stack>
      </div>
    );
  }

  private _onRenderItem = (props: ISelectedPeopleItemProps): JSX.Element => {
    return <ExtendedSelectedItem {...props} />;
  };

  private _onAddItemButtonClicked = (): void => {
    if (this._selectionList.current) {
      const { nextPersonIndex } = this.state;
      this._selectionList.current.addItems([people[nextPersonIndex]]);
      this.setState({ nextPersonIndex: nextPersonIndex + 1 });
    }
  };

  private _onExpandItem = (item: IExtendedPersonaProps): void => {
    const expandedItem =
      item.text === 'Group One' ? groupOne : item.text === 'Group Two' ? groupTwo : [];
    this._selectionList.current!.replaceItem(item, expandedItem);
  };

  private _onCopyItems(items: IExtendedPersonaProps[]): string {
    return items.map((item: IExtendedPersonaProps) => item.text).join(', ');
  }
}

const baseProductionCdnUrl =
  'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/';

const TestImages = {};

const people: (IExtendedPersonaProps & { key: string | number })[] = [];

const mru: IExtendedPersonaProps[] = people.slice(0, 5);

const groupOne: IExtendedPersonaProps[] = people.slice(6, 10);

const groupTwo: IExtendedPersonaProps[] = people.slice(11, 16);

ReactDOM.render(
  <Fabric>
    <SelectedPeopleListBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;

exports[`codepen transform handles examples importing TestImages 1`] = `
"const {
  DocumentCard,
  DocumentCardActivity,
  DocumentCardPreview,
  DocumentCardTitle,
  IDocumentCardPreviewProps,
  ImageFit,
  Fabric
} = window.Fabric;

class DocumentCardBasicExample extends React.PureComponent<{}> {
  public render(): JSX.Element {
    const previewProps: IDocumentCardPreviewProps = {
      previewImages: [
        {
          name: 'Revenue stream proposal fiscal year 2016 version02.pptx',
          linkProps: {
            href: 'http://bing.com'
          },
          previewImageSrc: TestImages.documentPreview,
          iconSrc: TestImages.iconPpt,
          imageFit: ImageFit.cover,
          width: 318,
          height: 196
        }
      ]
    };

    return (
      <DocumentCard onClickHref=\\"http://bing.com\\">
        <DocumentCardPreview {...previewProps} />
        <DocumentCardTitle
          title=\\"Large_file_name_with_underscores_used_to_separate_all_of_the_words_and_there_are_so_many_words_it_needs_truncating.pptx\\"
          shouldTruncate={true}
        />
        <DocumentCardActivity
          activity=\\"Created a few minutes ago\\"
          people={[{ name: 'Annie Lindqvist', profileImageSrc: TestImages.personaFemale }]}
        />
      </DocumentCard>
    );
  }
}

const baseProductionCdnUrl =
  'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/';

const TestImages = {};

ReactDOM.render(
  <Fabric>
    <DocumentCardBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;

exports[`codepen transform handles examples with class components 1`] = `
"const { SpinButton, Fabric } = window.Fabric;

class SpinButtonBasicExample extends React.Component<any, any> {
  public render(): JSX.Element {
    return (
      <div style={{ width: '400px' }}>
        <SpinButton
          defaultValue=\\"0\\"
          label={'Basic SpinButton:'}
          min={0}
          max={100}
          step={1}
          iconProps={{ iconName: 'IncreaseIndentLegacy' }}
          incrementButtonAriaLabel={'Increase value by 1'}
          decrementButtonAriaLabel={'Decrease value by 1'}
        />
      </div>
    );
  }
}

ReactDOM.render(
  <Fabric>
    <SpinButtonBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;

exports[`codepen transform handles examples with function components 1`] = `
"const { Label, Fabric } = window.Fabric;

const LabelBasicExample = () => {
  return (
    <div>
      <Label>I'm a Label</Label>
    </div>
  );
};

ReactDOM.render(
  <Fabric>
    <LabelBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;
